core/trino-spi/src/main/java/io/trino/spi/predicate/Utils.java

```diff
@@ -64,4 +69,42 @@ static RuntimeException handleThrowable(Throwable throwable)
         }
         return new RuntimeException(throwable);
     }
+
+    /// Try to expand given value set into a discrete set of objects.
+    /// For example: [1, 5] can be expanded into {1, 2, 3, 4, 5}.
+    public static Optional<Collection<Object>> tryExpandRanges(ValueSet valuesSet, int limit)
+    {
+        List<Range> ranges;
+        try {
+            ranges = valuesSet.getRanges().getOrderedRanges();
+        }
+        catch (UnsupportedOperationException e) {
+            return Optional.empty();
+        }
+        Type type = valuesSet.getType();
+        List<Object> result = new ArrayList<>(limit);
+        for (Range range : ranges) {
+            if (range.isLowUnbounded() || !range.isLowInclusive() || range.isHighUnbounded() || !range.isHighInclusive()) {
+                return Optional.empty();
```

Comment:
`! isLowInclusive` shouldn't trigger bailout, since you simply can (in pseudocode):

```
low = range.getLowBoundedValue();
if (! isLowInclusive) {
 low = type.getNearestGreaterValue(low);
 if (low.isEmpty()) { 
   bail out
 }
```